diff -r faddb1a6e933 libstdc++-v3/config/io/basic_file_stdio.cc
--- a/libstdc++-v3/config/io/basic_file_stdio.cc	Fri Jul 31 08:58:42 2009 -0600
+++ b/libstdc++-v3/config/io/basic_file_stdio.cc	Fri Jul 31 13:09:27 2009 -0600
@@ -236,4 +236,4 @@
-#ifdef _GLIBCXX_USE_LFS
+#if defined(_GLIBCXX_USE_LFS) || defined(__MINGW32__)
 	if ((_M_cfile = fopen64(__name, __c_mode)))
 #else
 	if ((_M_cfile = fopen(__name, __c_mode)))
@@ -317,7 +317,7 @@
   streamoff
   __basic_file<char>::seekoff(streamoff __off, ios_base::seekdir __way)
   {
-#ifdef _GLIBCXX_USE_LFS
+#if defined(_GLIBCXX_USE_LFS) || defined(__MINGW32__)
     return lseek64(this->fd(), __off, __way);
 #else
     if (__off > numeric_limits<off_t>::max()
@@ -366,6 +366,15 @@
 							   ios_base::cur);
 	return std::min(__off, streamoff(numeric_limits<streamsize>::max()));
       }
+#elif defined(__MINGW32__)
+    struct __stat64 __buffer;
+    const int __err = _fstat64(this->fd(), &__buffer);
+    if (!__err && _GLIBCXX_ISREG(__buffer.st_mode))
+      {
+	const streamoff __off = __buffer.st_size - lseek64(this->fd(), 0,
+							   ios_base::cur);
+	return std::min(__off, streamoff(numeric_limits<streamsize>::max()));
+      }
 #else
     struct stat __buffer;
     const int __err = fstat(this->fd(), &__buffer);
diff -r faddb1a6e933 libstdc++-v3/include/ext/stdio_sync_filebuf.h
--- a/libstdc++-v3/include/ext/stdio_sync_filebuf.h	Fri Jul 31 08:58:42 2009 -0600
+++ b/libstdc++-v3/include/ext/stdio_sync_filebuf.h	Fri Jul 31 13:09:27 2009 -0600
@@ -163,7 +163,7 @@
 	  __whence = SEEK_CUR;
 	else
 	  __whence = SEEK_END;
-#ifdef _GLIBCXX_USE_LFS
+#if defined(_GLIBCXX_USE_LFS) || defined(__MINGW32__)
 	if (!fseeko64(_M_file, __off, __whence))
 	  __ret = std::streampos(ftello64(_M_file));
 #else
